### PR TITLE
Support search by zero address.

### PR DESCRIPTION
**Pull request by tolikzinovyev**
## Summary

Zero address is a real address and we shouldn't ignore it. People are curious about transactions in which the zero address participates.

### CODE REVIEW

**Review by winder — APPROVED** [review]
This looks good, 2 requests for before you merge:

1. Rename the commit with the feature being added so that the release notes are easier to write: `Support search by zero address.`
2. Add comments for each switch case about why zero is only allowed sometimes. I'm pretty sure you got them all correct, but these are subtle details so things comments like "AssetSender is optional, it supposed to be zero unless used in a clawback transaction" may help the next person reading this code.

**Comment by tolikzinovyev** [comment]
Thanks, done. I wonder if we need to check for zero address in `buildTransactionQuery()`. For example, it doesn't make sense to set address to zero and address role to `AddressRoleCloseRemainderTo` in `TransactionFilter`. Should we trust the user to construct the right request?

**Comment by winder** [comment]
> Thanks, done. I wonder if we need to check for zero address in `buildTransactionQuery()`. For example, it doesn't make sense to set address to zero and address role to `AddressRoleCloseRemainderTo` in `TransactionFilter`. Should we trust the user to construct the right request?

In that case, we'd just return no results. Technically correct, right? I don't think we need to bother with custom error messages.

**Review by winder — APPROVED** [review]

**Comment by tolikzinovyev** [comment]
> > Thanks, done. I wonder if we need to check for zero address in `buildTransactionQuery()`. For example, it doesn't make sense to set address to zero and address role to `AddressRoleCloseRemainderTo` in `TransactionFilter`. Should we trust the user to construct the right request?
> 
> In that case, we'd just return no results. Technically correct, right? I don't think we need to bother with custom error messages.

No. Suppose the query is Address = zero, Role = CloseAddress. It will find a transaction where receiver is zero and close address is zero. But it will not find a transaction where receiver is non-zero and close address is zero. We could suppose the user should not make such a query. Any thoughts?

**Comment by winder** [comment]
> > > Thanks, done. I wonder if we need to check for zero address in `buildTransactionQuery()`. For example, it doesn't make sense to set address to zero and address role to `AddressRoleCloseRemainderTo` in `TransactionFilter`. Should we trust the user to construct the right request?
> > 
> > 
> > In that case, we'd just return no results. Technically correct, right? I don't think we need to bother with custom error messages.
> 
> No. Suppose the query is Address = zero, Role = CloseAddress. It will find a transaction where receiver is zero and close address is zero. But it will not find a transaction where receiver is non-zero and close address is zero. We could suppose the user should not make such a query. Any thoughts?

I'm pretty this is OK. We use the txn_participation table to identify the initial set of transactions, but then have a second filter on the jsonb to match the specific field we're looking for. I guess it might timeout if there are too many transactions li ke this for the slow jsonb filter.

**Comment by winder** [comment]
@tolikzinovyev I thought about this over the weekend and think the simplest solution is to prevent using AddressRole and a zero address. In `converter_utils.go : transactionParamsToTransactionFilter` this should do it:
```
	// Don't allow address role + zero address search, there are several invalid configurations.
	if params.AddressRole != nil && params.Address != nil {
		var zeroAddr basics.Address
		if zeroAddr.String() == *params.Address {
			errorArr = append(errorArr, errInvalidZeroAddressSearch)
		}
	}
```

**Review by winder — APPROVED** [review]
~LGTM~

File: api/handlers.go [hunk 1]
```diff
@@ -487,6 +487,15 @@ func (si *ServerImplementation) SearchForTransactions(ctx echo.Context, params g
...
 
+	if filter.AddressRole != 0 {
+		var address basics.Address
+		copy(address[:], filter.Address)
+		if address.IsZero() {
+			return badRequest(
+				ctx, "searching transactions by zero address with address role is not supported")
+		}
+	}
+
```
Comment (winder) [inline api/handlers.go:498]: This should go into `transactionParamsToTransactionFilter` or `si.fetchTransactions` to make sure the constraint is checked for all account searches.

**Review by winder — CHANGES_REQUESTED** [review]
I think the parameter checking should go in a different spot

File: api/handlers.go (hunk 1 above)
Comment (tolikzinovyev) [reply api/handlers.go:498]: Only `LookupApplicationLogsByID()` would violate, which I think is fine. If we do put it in `fetchTransactions()`, `LookupApplicationLogsByID()` would return a weird error.

Comment (tolikzinovyev) [reply api/handlers.go:498]: `transactionParamsToTransactionFilter()` is a conversion function, I don't want to put business logic there.

I think the current code is ok, especially given that no one in their right mind will use `LookupApplicationLogsByID()` with zero address.

Comment (winder) [reply api/handlers.go:498]: It also validates the input, there's a test in `api/handlers_test.go` for it also so you can add a test:
```
    {
      name: "Invalid zero address parameters",
      params: generated.SearchForTransactionsParams{
        Address:     strPtr(basics.Address{}.String()),
        AddressRole: 9,
      },
      filter:        idb.TransactionFilter{},
      errorContains: []string{errInvalidZeroAddressSearch},
    }
```

Could you also add the error message to `error_messages.go`?

Comment (tolikzinovyev) [reply api/handlers.go:497]: Moved the error message to `error_messages.go`. `transactionParamsToTransactionFilter()` does a couple of sanity checks, but no more. Definitely no artificial limitations should be in there.

Comment (tolikzinovyev) [reply api/handlers.go:545]: I think it is incorrect that https://github.com/algorand/indexer/blob/830e1ba83500b2c78251070fa67031efdd418ed7/api/handlers.go#L277 becomes `idb.AddressRoleSender | idb.AddressRoleAssetSender`, so I set `filter.AddressRole` directly and run validation that used to be in `transactionParamsToTransactionFilter()` after.

File: api/handlers.go [hunk 2]
```diff
@@ -274,13 +315,18 @@ func (si *ServerImplementation) LookupApplicationLogsByID(ctx echo.Context, appl
...
 		MaxRound:      params.MaxRound,
 		Address:       params.SenderAddress,
-		AddressRole:   strPtr(addrRoleSender),
 	}
 
 	filter, err := transactionParamsToTransactionFilter(searchParams)
 	if err != nil {
 		return badRequest(ctx, err.Error())
 	}
+	filter.AddressRole = idb.AddressRoleSender
```
Comment (winder) [inline api/handlers.go:324]: Nice :+1:

**Review by winder — APPROVED** [review]
LGTM

**Pull request merged**